server/src/main/java/com/paypal/selion/proxy/SeLionSauceProxy.java

```diff
@@ -22,72 +22,248 @@
+        }
+    }
+
+    private void processParentTunnelCapability(Map<String, Object> requestedCapabilities) {
+        final String tunnelId = (String) requestedCapabilities.get(SauceLabsCapability.PARENT_TUNNEL_CAPABILITY);
+        if (StringUtils.isBlank(tunnelId) && StringUtils.isNotBlank(defaultParentTunnel)) {
+            // apply the default parent-tunnel according to the configured policy
+            final String actualUserName = (String) requestedCapabilities.get(SauceLabsCapability.USERNAME_CAPABILITY);
+            requestedCapabilities.put(SauceLabsCapability.PARENT_TUNNEL_CAPABILITY, (defaultParentTunnel
+                    .equalsIgnoreCase(DEFAULT_PARENT_TUNNEL_TO_USERNAME_TOKEN)) ? actualUserName : defaultParentTunnel);
+        }
+    }
+
+    private void processTunnelIdentifierCapability(Map<String, Object> requestedCapabilities) {
+        final String tunnelId = (String) requestedCapabilities.get(SauceLabsCapability.TUNNEL_IDENTIFIER_CAPABILITY);
+        if (StringUtils.isBlank(tunnelId) && StringUtils.isNotBlank(defaultTunnelIdentifier)) {
+            requestedCapabilities.put(SauceLabsCapability.TUNNEL_IDENTIFIER_CAPABILITY, defaultTunnelIdentifier);
+        }
+    }
+
+    private void processCredentails(Map<String, Object> requestedCapabilities) {
+        if (sauceConfigReader.isRequireUserCredentials()) {
+            return;
+        }
+
+        final String username = (String) requestedCapabilities.get(SauceLabsCapability.USERNAME_CAPABILITY);
+        final String accessKey = (String) requestedCapabilities.get(SauceLabsCapability.ACCESS_KEY_CAPABILITY);
+        if (StringUtils.isBlank(username)) {
+            requestedCapabilities.put(SauceLabsCapability.USERNAME_CAPABILITY, sauceConfigReader.getUserName());
+        }
+        if (StringUtils.isBlank(accessKey)) {
+            requestedCapabilities.put(SauceLabsCapability.ACCESS_KEY_CAPABILITY, sauceConfigReader.getApiKey());
+        }
+    }
+
+    private boolean isMissingRequiredCapabilities(Map<String, Object> requestedCapabilities) {
+        return requestedCapabilities.get(SauceLabsCapability.USERNAME_CAPABILITY) == null
+                || requestedCapabilities.get(SauceLabsCapability.ACCESS_KEY_CAPABILITY) == null;
+    }
+
+    @Override
+    public void beforeCommand(TestSession session, HttpServletRequest request, HttpServletResponse response) {
+        Map<String, Object> requestedCapabilities = session.getRequestedCapabilities();
+
+        // error the session when there is no sauce user name and/or access key
+        if (isMissingRequiredCapabilities(requestedCapabilities)) {
+            throw new GridException("Sauce Labs credentials were not specified.");
+        }
+
+        // give up if sauce labs is down ...
+        // this should fail faster vs. returning the session to hub by doing this check in #getNewSession
+        // (DefaultRemoteProxy's behavior)
```

Comment:
code does not currently have desired effect.. the current code is depending on the check in `DefaultRemoteProxy` which if `down` will kick the new session request back to the hub.. making this block to not likel be reached.